modules/core/src/com/haulmont/cuba/security/app/EntityAttributeChanges.java

```diff
@@ -14,38 +14,71 @@
  * limitations under the License.
  */
 
-package com.haulmont.cuba.core.sys.persistence;
+package com.haulmont.cuba.security.app;
 
+import com.haulmont.cuba.core.entity.Entity;
+import org.eclipse.persistence.descriptors.changetracking.ChangeTracker;
+import org.eclipse.persistence.internal.descriptors.changetracking.AttributeChangeListener;
 import org.eclipse.persistence.sessions.changesets.AggregateChangeRecord;
 import org.eclipse.persistence.sessions.changesets.ChangeRecord;
 import org.eclipse.persistence.sessions.changesets.ObjectChangeSet;
 
 import javax.annotation.Nullable;
-import java.util.*;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
 import java.util.stream.Collectors;
-import java.util.stream.Stream;
 
 /**
- * INTERNAL. Accumulates changes in entity attributes.
+ * Accumulates changes in entity attributes.
+ * Entity changes for entity is created by:
+ * <pre>
+ *  EntityAttributeChanges changes = new EntityAttributeChanges();
+ *  changes.addChanges(entity)
+ * </pre>
+ * System allows you to add changes for attributes manually by calling
+ * {@link EntityAttributeChanges#addChange(String, Object)}
  */
 public class EntityAttributeChanges {
 
-    private Set<Change> changes = new HashSet<>();
-    private Map<String, EntityAttributeChanges> embeddedChanges = new HashMap<>();
+    protected Set<Change> changes = new HashSet<>();
+    protected Map<String, EntityAttributeChanges> embeddedChanges = new HashMap<>();
 
-    public void addChanges(ObjectChangeSet changeSet) {
-        if (changeSet == null)
+    /**
+     * Add change for the entity attribute
+     *
+     * @param attributeName - attribute name
+     * @param oldValue      - old attribute value
+     */
+    public void addChange(String attributeName, Object oldValue) {
+        changes.add(new Change(attributeName, oldValue));
+    }
+
+    /**
+     * Add changes for the embedded entity attribute
+     *
+     * @param attributeName - attribute name
+     * @param changes       - {@link EntityAttributeChanges} changes for embedded object
+     */
+    public void addEmbeddedChanges(String attributeName, EntityAttributeChanges changes) {
+        embeddedChanges.put(attributeName, changes);
+    }
+
+    /**
+     * Accumulates changes for entity. Stores changed attribute name and old value
```

Comment:
Accumulates changes for the entity. Stores changed attribute names and old values.